src/middlewared/middlewared/plugins/pool_/dataset_details.py

```diff
@@ -0,0 +1,221 @@
+        rsynctasks = self.middleware.call_sync('rsynctask.query')
+        collapsed = []
+        info = self.middleware.call_sync('pool.dataset.query', filters, options)
+        for dataset in info:
+            self.collapse_datasets(dataset, collapsed)
+
+        for i in collapsed:
+            snapshot_count, locked = self.get_snapcount_and_encryption_status(i, mntinfo)
+            i['snapshot_count'] = snapshot_count
+            i['locked'] = locked
+            i['nfs_shares'] = self.get_nfs_shares(i, nfsshares, mntinfo)
+            i['smb_shares'] = self.get_smb_shares(i, smbshares, mntinfo)
+            i['iscsi_shares'] = self.get_iscsi_shares(i, iscsishares, mntinfo)
+            i['replication_tasks_count'] = self.get_repl_tasks_count(i, repltasks)
+            i['snapshot_tasks_count'] = self.get_snapshot_tasks_count(i, snaptasks)
+            i['cloudsync_tasks_count'] = self.get_cloudsync_tasks_count(i, cldtasks)
+            i['rsync_tasks_count'] = self.get_rsync_tasks_count(i, rsynctasks)
+
+        return info
+
+    @private
+    def collapse_datasets(self, dataset, collapsed):
+        collapsed.append(dataset)
+        for child in dataset.get('children', []):
+            self.collapse_datasets(child, collapsed)
+
+    @private
+    def build_iscsi_share_info(self):
+        t_to_e = self.middleware.call_sync('iscsi.targetextent.query')
+        t = {i['id']: i for i in self.middleware.call_sync('iscsi.target.query')}
+        e = {i['id']: i for i in self.middleware.call_sync('iscsi.extent.query')}
+
+        iscsi_shares = []
+        for i in filter(lambda x: x['target'] in t and t[x['target']]['groups'] and x['extent'] in e, t_to_e):
+            """
+            1. make sure target's and extent's id exist in the target to extent table
+            2. make sure the target has `groups` entry since, without it, it's impossible
+                that it's being shared via iscsi
+            """
+            iscsi_shares.append({'extent': e[i['extent']], 'target': t[i['target']]})
+
+        return iscsi_shares
+
+    @private
+    def get_snapcount_and_encryption_status(self, ds, mntinfo):
+        snap_count = 0
+        locked = False
+        if ds['type'] == 'FILESYSTEM':
+            # FIXME: determining zvol snapshot count requires iterating over
+            # all snapshots for a given zvol and then counting them which is
+            # painful. This will be moot when this is merged upstream
+            # https://github.com/openzfs/zfs/pull/13635
+
+            try:
+                st = os.stat(f'{ds["mountpoint"]}/.zfs/snapshot')
+            except FileNotFoundError:
+                # means that the zfs snapshot dir doesn't exist which
+                # can only happen if the dataset is encrypted and "locked"
+                # (unmounted) or the dataset is unmounted
+                locked = ds['encrypted']
+            else:
+                if st.st_ino == ZFSCTL.INO_SNAPDIR:
```

Comment:
`pool.dataset.snapshot_count` uses the same logic, let's have it in only one place please.

Reply:
@themylogin I didn't use `snapshot_count` because it will run `zfs.snapshot.query` if it can't find the `/.zfs/snapshot` directory which makes this entire method unusable. That's why I said I'm not querying zvols right now.

Reply:
We can extract 
```
                if stat.st_ino == ZFSCTL.INO_SNAPDIR:
                    return stat.st_nlink - 2
```
into a separate method like `snapshot_count_for_stat`